### PR TITLE
Add uptraceexporter

### PR DESCRIPTION
**Pull request by vmihailenco**
**Description:**
Allow exporting trace data to Uptrace.dev

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2021

**Testing:**
Unit tests are included

**Documentation:**
The readme includes a short description and the configuration.

**Notes**
- This PR uses `DSN` (AKA connection string) instead of `endpoint`. They look very similar but are rather different, i.e. endpoint is constructed from `DSN`. [wiki](https://en.wikipedia.org/wiki/Data_source_name#:~:text=In%20computing%2C%20a%20data%20source,connection%20to%20a%20data%20source.)
- This is the first of 3 PRs. The second will include updates to the code owners and dependabot, and the third will add it to the binary.

### CODE REVIEW

File: exporter/uptraceexporter/config.go
Comment (vmihailenco) [reply exporter/uptraceexporter/config.go:1]: Okay, will update to 2021

File: exporter/uptraceexporter/exporter.go
Comment (vmihailenco) [reply exporter/uptraceexporter/exporter.go:68]: AFAIK that this removes lots of allocations. This way I have a slice of structs that are allocated in batches (https://www.darkcoding.net/software/go-how-slices-grow/).

>Why aren't you first building the span (out) and then appending to the list of spans?

Then I need to allocate a separate struct and then copy it to a slice. That is an allocation for each struct. Or using a value instead of a pointer and use more CPU for copying.

File: exporter/uptraceexporter/exporter.go
Comment (vmihailenco) [reply exporter/uptraceexporter/exporter.go:96]: Could you point me to an exporter that implements this? Or explain this in a more detail?

File: exporter/uptraceexporter/exporter_test.go
Comment (vmihailenco) [reply exporter/uptraceexporter/exporter_test.go:47]: Tests don't make network requests. This is needed to pass DSN validation and can be something different, e.g. `http://key@domain.com/1`

File: exporter/uptraceexporter/exporter.go
Comment (jpkrohling) [reply exporter/uptraceexporter/exporter.go:68]: Didn't know about that, thanks!

File: exporter/uptraceexporter/exporter.go
Comment (jpkrohling) [reply exporter/uptraceexporter/exporter.go:96]: Sure, here's an example that records two things: latency of the requests and success/errors

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/74dae1e13c8476eeb42cdd5d1110cb8363d5d2fe/exporter/loadbalancingexporter/exporter.go#L220-L231

Only one `record` call is made, and exposed as different views:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/74dae1e13c8476eeb42cdd5d1110cb8363d5d2fe/exporter/loadbalancingexporter/metrics.go#L60-L78

File: exporter/uptraceexporter/exporter_test.go
Comment (jpkrohling) [reply exporter/uptraceexporter/exporter_test.go:47]: Sounds good then

**Comment by vmihailenco** [comment]
@jpkrohling thanks for the review. I've added backend latency metric based on your example and code from [otel-go-contrib](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/c1c564f356718cadf14764d8d0c584ff2f46a406/instrumentation/net/http/otelhttp/handler.go#L105-L106). Please take another look.

I see that codecov check is failing but it is complaining about few branches in utility functions like

```go
func statusCode(status pdata.SpanStatus) string {
	switch status.Code() {
	case pdata.StatusCodeOk:
		return "ok"
	case pdata.StatusCodeError:
		return "error"
	}
	return "unset" // not covered
}
```

Hopefully I don't have to cover all of them :)

File: exporter/uptraceexporter/exporter.go
Comment (vmihailenco) [inline exporter/uptraceexporter/exporter.go:50]: Do I really need to use such long name? :confused:

File: exporter/uptraceexporter/exporter.go
Comment (vmihailenco) [inline exporter/uptraceexporter/exporter.go:128]: I am not too happy with this but I did not find a better way to check if `meter.NewInt64ValueRecorder` returned an error or not...

File: exporter/uptraceexporter/exporter.go
Comment (vmihailenco) [reply exporter/uptraceexporter/exporter.go]: Done but using Otel Metrics and without any views. Hopefully it is enough :)

File: exporter/uptraceexporter/exporter.go
Comment (jpkrohling) [reply exporter/uptraceexporter/exporter.go:50]: I don't think we need. I wasn't aware that we should be using the OpenTelemetry API for this already, but for the legacy OpenCensus one this wasn't necessary:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9f137ea75d249631c411b9d845e83a179032a5b8/exporter/loadbalancingexporter/metrics.go#L24

And then:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9f137ea75d249631c411b9d845e83a179032a5b8/exporter/loadbalancingexporter/resolver_dns.go#L125

Where did you see this usage?

Comment (vmihailenco) [reply exporter/uptraceexporter/exporter.go:50]: I try to follow docs at https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md#obtaining-a-meter which mentions "io.opentelemetry.contrib.mongodb" for Java.

>I wasn't aware that we should be using the OpenTelemetry API for this already

Yeah, I guess metrics are not stable yet. But using OpenCensus makes even less sense (to me)...

What do you think if I just drop the metrics instrumentation for now? Add add them in a separate PR later when Otel Metrics are ready?

Comment (jpkrohling) [reply exporter/uptraceexporter/exporter.go:50]: I know your feeling, I had it too. But it's either that, or you won't have metrics for your component, which is even worse for you  and for your users. Besides, all other components are instrumented with OpenCensus at the moment and will likely be mass-migrated once we are ready with OpenTelemetry Metrics.

Comment (vmihailenco) [reply exporter/uptraceexporter/exporter.go:50]: >which is even worse for you and for your users

But I have tracing :) SendSpans is instrumented with Otel traces so I have even more data than with metrics. Besides Uptrace does not support metrics at the moment...

>are instrumented with OpenCensus

Just to make it clear  - should I replace Otel Metrics with OpenCensus for now?

Comment (jpkrohling) [reply exporter/uptraceexporter/exporter.go:50]: > Besides Uptrace does not support metrics at the moment

This is about understanding what the collector is doing. If a user of yours is using the collector and see no data in your service, how do you intend to figure out what's going on? The collector exposes a `/metrics` by default, which includes data generated by the individual components. If you think you have enough information elsewhere to support your users, that's fine, you can skip this.

> Just to make it clear - should I replace Otel Metrics with OpenCensus for now?

If you think you should have metrics to understand what your exporter is doing, yes. If you decide to not have metrics at all, I'm also fine.

Comment (vmihailenco) [reply exporter/uptraceexporter/exporter.go:50]: If this is not a hard requirement I would prefer to add Otel metrics later in a separate PR. Few other exporters I've checked don't have metrics too :)

I don't have experience with OpenCensus and learning it at this stage does not look worth it even if basic concepts are the same...

Comment (jpkrohling) [reply exporter/uptraceexporter/exporter.go:50]: Alright, feel free to update the PR to remove the metric.

**Comment by jpkrohling** [comment]
If it's not technically feasible to improve the code coverage, we can waive the requirement, but it doesn't look like it's the case. So, please try to please the codecov deity :-)

**Comment by vmihailenco** [comment]
@jpkrohling I've removed metrics for now and bumped code coverage. PTAL

File: exporter/uptraceexporter/config_test.go
Comment (jpkrohling) [inline exporter/uptraceexporter/config_test.go:41]: For reference, you could have used `require.Len` here.

**Review by jpkrohling — DISMISSED** [review]

File: exporter/uptraceexporter/config_test.go
Comment (vmihailenco) [reply exporter/uptraceexporter/config_test.go:41]: Ack, but I don't want to trigger the build again - I had to re-run it few times before the build passed (some unrelated failures :)

Comment (jpkrohling) [reply exporter/uptraceexporter/config_test.go:41]: That's fine, it's approved as is :-)

Comment (vmihailenco) [reply exporter/uptraceexporter/config_test.go:41]: Thanks for the review :)

File: exporter/uptraceexporter/config.go
Comment (bogdandrutu) [inline exporter/uptraceexporter/config.go:24]: We use endpoint for all the configs, is there anything special here that you need to name it `dns`?

**Review by bogdandrutu — COMMENTED** [review]
Please follow the instructions about adding a new component https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component

File: exporter/uptraceexporter/config.go
Comment (jpkrohling) [reply exporter/uptraceexporter/config.go:24]: Endpoint kinds implies the format that is used by the other receivers, like `hostname:port`. If this property is changed to be named `endpoint`, it should follow the same convention.

Comment (bogdandrutu) [reply exporter/uptraceexporter/config.go:24]: @jpkrohling good call, but I would like to understand if this exporter only works with a DNS address or we can have a simple endpoint that can also be a DNS entry if needed.

Comment (jpkrohling) [reply exporter/uptraceexporter/config.go:24]: Note that DSN != DNS. A DSN (data source name) is typically a URL-like string, and is used typically for database connection strings.

For this here, the DSN should be broken down into different components, like we have for other exporters. Also, the HTTP Client should also be built like the others, to take advantage of common facilities. Not sure how I could have missed this!

Comment (vmihailenco) [reply exporter/uptraceexporter/config.go:24]: >is there anything special here that you need to name it dns?

It is a DSN or a connection string. Is there anything wrong it? Is it banned in Otel collector?

>For this here, the DSN should be broken down into different components

What components? How this makes anyone's life better or simpler?

Uptrace UI provides single string to control access. How breaking that single string into multiple components for Uptrace exporter in Otel collector makes any sense or improves anything?

>We use endpoint for all the configs
> like we have for other exporters

What do you mean by this? They all provide different ways to configure their exporters...

- honeycomb has `api_key`, `dataset`, `api_url`
- jaeger has `url`, `timeout`, `headers`
- newrelic has `spans_url_override`, `timeout`, `apikey`

>Also, the HTTP Client should also be built like the others

Do you care to point me to an example in Honeycomb/Newrelic/Jaeger exporters? Or some documentation?

**Comment by vmihailenco** [comment]
>Please follow the instructions about adding a new component https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component

That is exactly what I've followed. It is not obvious what you mean.

File: exporter/uptraceexporter/config.go
Comment (jpkrohling) [reply exporter/uptraceexporter/config.go:24]: > What components? How this makes anyone's life better or simpler?

Protocol, hostname+port. I assume those are the only things that might change. For users of the OpenTelemetry Collector, having your exporter to follow what other exporters are doing is making their lives easier.

> Uptrace UI provides single string to control access. How breaking that single string into multiple components for Uptrace exporter in Otel collector makes any sense or improves anything?

If your users are used to this concept, then I'm certainly fine in letting you use it.

> What do you mean by this? All exporters provide different configurations keys...

I can only speak for Jaeger, and we have only the `endpoint` to control the destination. Not sure which URL you are talking about.

https://github.com/open-telemetry/opentelemetry-collector/blob/68c5961f7bc2d8d55fd431ab07367fc16f8b26d0/exporter/jaegerexporter/config.go#L24-L31

https://github.com/open-telemetry/opentelemetry-collector/blob/68c5961f7bc2d8d55fd431ab07367fc16f8b26d0/config/configgrpc/configgrpc.go#L67

> Do you care to point me to an example in Honeycomb/Newrelic/Jaeger exporters? Or some documentation?

Again, I can only speak for Jaeger, but here's how Jaeger builds a gRPC client connection:

https://github.com/open-telemetry/opentelemetry-collector/blob/68c5961f7bc2d8d55fd431ab07367fc16f8b26d0/exporter/jaegerexporter/exporter.go#L43-L51

It uses the common gRPC configuration object, so that every new property added to that common config will automatically be available to the Jaeger exporter. Similarly, all the TLS configuration is handled by the common config object, providing a single experience to the users.

In my opinion, it's ultimately your decision how you want your exporter to look like, but it should feel like any other exporter whenever possible.

Comment (vmihailenco) [reply exporter/uptraceexporter/config.go:24]: > Not sure which URL you are talking about.

I am talking about [this](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/jaegerthrifthttpexporter/testdata/config.yaml#L8-L15) and it looks rather different even though  it is also Jaeger exporter...

Comment (jpkrohling) [reply exporter/uptraceexporter/config.go:24]: True. To be fair, that's quite old and would have been written following the current standards. Good thing is that this isn't the recommended exporter :-)

**Comment by jpkrohling** [comment]
@vmihailenco, are you interested in adjusting this PR?

**Comment by vmihailenco** [comment]
@jpkrohling you've dismissed your approval because of `DSN`. But at the same time it looks like you don't mind keeping it and I can't guess what @bogdandrutu thinks. I personally don't see how `DSN` is better than `endpoint`. And Uptrace already uses `DSN` everywhere and that is what users are familiar with.

Another thing that was mentioned is common TLS / HTTP client configuration. There are 2 problems with that. First I am re-using code from https://github.com/uptrace/uptrace-go/tree/master/spanexp which already works and has a HTTP client. I guess with some refactoring I can make this PR honor common TLS configuration (which really should be called custom configuration).

But does it make sense? I understand that we **need** custom TLS configuration for hosted Otel collector, because we must provide TLS certificates and often people use self-signed certificates. That is not the case with Uptrace. The default TLS configuration should always work. Actually it even makes sense to have 2 different custom TLS configs - one for Uptrace and one for OTLP exporter if you want to export to another hosted Otel collector.

I could also mention that other exporters don't follow these guidelines. You could say that they are outdated, but I am saying that this is what they actually want/need. Thus the question - whose life I am making simpler by renaming DSN or re-using custom TLS config provided by Otel collector? Definitely not mine. And I fail to see how it makes a change for you.

So what should I adjust?

**Comment by jpkrohling** [comment]
I see your arguments, but in the end, there's an intersection of users who are both OpenTelemetry Collector users and uptrace users. They might be used to either OpenTelemetry's standards, or to uptrace standards. In general, we should strive to make users comfortable with both worlds, using the terms they know. 

Most exporters out there are trying to follow the common helpers, providing users with a consistent way to configure components: same property names, same semantics, ... Sure, there are some that deviate from the suggestions, likely because they are old, and there are cases where the differences can't be reasonably reconciled, like in the `elasticapmexporter` (IIRC), where their `endpoint` is typically multiple endpoints. And even in those cases, the same features are provided under the same configuration property names, following the same semantics.

In my view, your HTTP client options could be replaced by the helpers with minimal effort, and you'd provide all the other features available to other exporters "for free", in a consistent way (retry with backoff, sending queue, timeouts, ...)

If you are saying that uptrace customers using the OpenTelemetry Collector are better served by deviating from the OpenTelemetry Collector's conventions, then so be it. I'm happy with approving the PR if you believe so, but a maintainer would also have to approve this before merging.

**Comment by vmihailenco** [comment]
>are better served by deviating from the OpenTelemetry Collector's conventions

I was trying to say that I can't imagine users configuring those options.

>In my view, your HTTP client options could be replaced by the helpers with minimal effort, and you'd provide all the other features available to other exporters "for free", in a consistent way (retry with backoff, sending queue, timeouts, ...)

Thanks for the suggestion. I believe I've just done that in a separate commit. I've copied documentation for those "new" config options from OTLP exporter.

Please take another look. Do you think I can keep DSN? It is quite different from `endpoint` and is widely used, e.g. Sentry uses it.

**Comment by vmihailenco** [comment]
@jpkrohling gentle ping in case you missed this. I've also added a note on DSN to the PR description.

File: exporter/uptraceexporter/exporter_test.go
Comment (jpkrohling) [reply exporter/uptraceexporter/exporter_test.go:47]: Are you changing the domain to something else? `example.com` would be an RFC-compliant domain for this :-)

File: exporter/uptraceexporter/factory.go
Comment (jpkrohling) [inline exporter/uptraceexporter/factory.go:47]: You might want to add a warning to the logs in case users set this value.

**Review by jpkrohling — COMMENTED** [review]
Looks good, just a couple of minor comments.

File: exporter/uptraceexporter/factory.go
Comment (vmihailenco) [reply exporter/uptraceexporter/factory.go:47]: Done

File: exporter/uptraceexporter/exporter_test.go
Comment (vmihailenco) [reply exporter/uptraceexporter/exporter_test.go:55]: I've changed the domain to `example.com` (if I understand you correctly)

**Comment by vmihailenco** [comment]
@jpkrohling both done. PTAL

**Review by jpkrohling — APPROVED** [review]

**Comment by bogdandrutu** [comment]
@vmihailenco 

>> Please follow the instructions about adding a new component https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component

>That is exactly what I've followed. It is not obvious what you mean.

Maybe is not clear in that document, or maybe you ignored it, but I would not expect code that translates to the exporter format in the first PR.

```
First PR should include the overall structure of the new component:
Readme, configuration, and factory implementation usually using the helper factory structs.
This PR is usually trivial to review, so the size limit does not apply to it.
```

**Comment by vmihailenco** [comment]
@bogdandrutu I've removed the code you've mentioned from this PR. PTAL

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
